autoemulate/experimental/calibration/hmc.py

```diff
@@ -0,0 +1,238 @@
+            A dictionary mapping input parameter names to their (min, max) ranges.
+        observations: dict[str, TensorLike]
+            A dictionary of observations for each output.
+        observation_noise: float |  dict[str, float]
+            A single value or a dictionary of values (one per output). Defaults to 0.1.
+        calibration_params: list[str] | None
+            Optional list of input parameters to calibrate. Any parameters that are not
+            listed will be set to the midpoint value of their parameter range. If None,
+            will calibrate all input parameters. Defaults to None.
+        device: DeviceLike | None
+            The device to use. If None, the default torch device is returned.
+            TODO: do we need to do anything more to ensure the device is correctly
+            handled for the pyro model?
+
+        Notes
+        -----
+        The model assumes:
+        - Uniform priors for calibrated parameters (bounds given by `parameters_range`)
+        - Gaussian likelihood with no correlation between outputs
+        All non-calibrated parameters are set to a constant value. This is chosen as the
+        midpoint value of `parameters_range`.
+        """
+        TorchDeviceMixin.__init__(self, device=device)
+        self.parameter_range = parameter_range
+        if calibration_params is None:
+            calibration_params = list(parameter_range.keys())
+        self.calibration_params = calibration_params
+        self.emulator = emulator
+        self.output_names = list(observations.keys())
+
+        # Check observation tensors are 1D (convert if 0D)
+        processed_observations = {}
+        obs_lengths = []
+        for output, obs in observations.items():
+            if obs.ndim == 0:
+                corrected_obs = obs.unsqueeze(0)
+            elif obs.ndim > 1:
+                raise ValueError(f"Tensor for output '{output}' is not 1D.")
+            else:
+                corrected_obs = obs
+            processed_observations[output] = corrected_obs
+            obs_lengths.append(corrected_obs.shape[0])
+        if len(set(obs_lengths)) != 1:
+            msg = "All outputs must have the same number of observations."
+            raise ValueError(msg)
+        self.observations = processed_observations
+        self.n_observations = obs_lengths[0]
+
+        # Save observation noise as {output: value} dictionary
+        if isinstance(observation_noise, float):
+            self.observation_noise = dict.fromkeys(self.output_names, observation_noise)
+        elif isinstance(observation_noise, dict):
+            self.observation_noise = observation_noise
+        else:
+            msg = "Noise must be either a float or a dictionary of floats."
+            raise ValueError(msg)
+
+    def model(self, predict: bool = False):
+        """
+        Pyro model.
+
+        Parameters
+        ----------
+        predict: bool
```

Comment:
The description here isn't correct. When the model is run with `predict=True` we aren't actually sampling. The posterior predictive is returned from the generated parameter MCMC samples.